azure/services/virtualmachineimages/images.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	if osAndVersion == "" {
+		osAndVersion = azure.DefaultWindowsOsAndVersion
+	}
+
+	publisher, offer := azure.DefaultImagePublisherID, azure.DefaultWindowsImageOfferID
+	skuID, version, err := getDefaultImageSKUIDAndVersion(
+		ctx, auth, location, publisher, offer, k8sVersion, osAndVersion)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get default image")
+	}
+
+	// Starting with 1.22 we default to containerd for Windows unless the runtime flag is set.
+	if v.GTE(v122) && runtime != "dockershim" {
+		skuID += "-containerd"
+	}
+
+	defaultImage := &infrav1.Image{
+		Marketplace: &infrav1.AzureMarketplaceImage{
+			Publisher: publisher,
+			Offer:     offer,
+			SKU:       skuID,
+			Version:   version,
+		},
+	}
+
+	return defaultImage, nil
+}
+
+// k8sVersionInSKUName returns true if the k8s version is in the SKU name (the older style of naming).
+func k8sVersionInSKUName(major, minor, patch uint64) bool {
+	return (major == 1 && minor < 21) ||
+		(major == 1 && minor == 21 && patch <= 12) ||
+		(major == 1 && minor == 22 && patch <= 9) ||
+		(major == 1 && minor == 23 && patch <= 6)
+}
+
+// GetDefaultImageSKUID gets the SKU ID and version of the image to use for the provided version of Kubernetes.
+// note: osAndVersion is expected to be in the format of {os}-{version} (ex: ubuntu-2004 or windows-2022)
+func getDefaultImageSKUIDAndVersion(ctx context.Context, auth azure.Authorizer, location, publisher, offer, k8sVersion, osAndVersion string) (string, string, error) {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "virtualmachineimages.AzureClient.getDefaultImageSKUIDAndVersion")
+	defer done()
+
+	v, err := semver.ParseTolerant(k8sVersion)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "unable to parse Kubernetes version \"%s\" in spec, expected valid SemVer string", k8sVersion)
+	}
+
+	// Old SKUs before 1.21.12, 1.22.9, or 1.23.6 are named like "k8s-1dot21dot2-ubuntu-2004".
+	if k8sVersionInSKUName(v.Major, v.Minor, v.Patch) {
+		return fmt.Sprintf("k8s-%ddot%ddot%d-%s", v.Major, v.Minor, v.Patch, osAndVersion), azure.LatestVersion, nil
+	}
+
+	// New SKUs don't contain the Kubernetes version and are named like "ubuntu-2004-gen1".
+	sku := fmt.Sprintf("%s-gen1", osAndVersion)
+	// Look up the version ID for the given SKU.
+	listVMImagesResource, err := NewClient(auth).List(ctx, location, publisher, offer, sku)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "unable to list VM images for publisher \"%s\" offer \"%s\" sku \"%s\"", publisher, offer, sku)
+	}
+	vmImages := listVMImagesResource.Value
+	if vmImages == nil || len(*vmImages) == 0 {
+		return "", "", errors.Errorf("no VM images found for publisher \"%s\" offer \"%s\" sku \"%s\"", publisher, offer, sku)
+	}
+
+	// sort the VM image names
+	names := []string{}
+	for _, vmImage := range *vmImages {
+		names = append(names, *vmImage.Name)
+	}
+	sort.Strings(names)
+
+	// Pick the first one whose version matches
```

Comment:
we need logic to get the latest version of a patch (by date)

Reply:
Since we've already alpha-sorted by name, picking the first one that matches the version prefix should give us the latest by date. Maybe I need to sort descending... But either way, I am working on a unit test to validate this part of the behavior, so stay tuned. :-)

Reply:
I needed to reverse the sorting, and now there are unit tests covering this.